g3doc/Tutorials/Toy/Ch-2.md

```diff
@@ -23,139 +23,320 @@ addresses this issue by being designed for extensibility. As such, there are
+components: attributes, operations, types, etc. This allows MLIR to parse,
+represent, and round-trip any valid IR. For example, we could inject our toy
+operation from above into an .mlir file and round-trip through *mlir-opt*
+without registering anything:
 
-```MLIR(.mlir)
-func @some_func(%arg0: !random_dialect<"custom_type">) -> !another_dialect<"other_type"> {
-  %result = "custom.operation"(%arg0) { attr = #random_dialect<"custom_attribute"> } : (!random_dialect<"custom_type">) -> !another_dialect<"other_type">
-  return %result : !another_dialect<"other_type">
+```mlir
+func @toy_func(%tensor: tensor<2x3xf64>) -> tensor<3x2xf64> {
+  %t_tensor = "toy.transpose"(%tensor) { inplace = true } : (tensor<2x3xf64>) -> tensor<3x2xf64>
+  return %t_tensor : tensor<3x2xf64>
 }
 ```
 
-Here MLIR will enforce some structural constraints (SSA, block termination,
-etc.) but otherwise the types and the `custom.operation` are completely opaque.
+In the cases of unregistered attributes, operations, types, MLIR will enforce
+some structural constraints (SSA, block termination, etc.) but otherwise they
+are completely opaque. This can be useful for bootstrapping purposes, but it is
+generally advised against. Opaque operations must be treated conservatively by
+transformations and analyses, cannot be properly verified for custom semantics,
+and are much harder to construct and manipulate.
 
-We will take advantage of this facility for the initial emission of MLIR for Toy
-by traversing the AST. Our operation names will be prefixed `toy.` in
-preparation for a `toy` dialect, which we will introduce with more details in
-the [next chapter](Ch-3.md).
+## Defining a Toy Dialect
 
-Programmatically creating an opaque operation, like the one above, involves
-using the `mlir::OperationState` structure which group all the basic elements
-needed to build an operation with an `mlir::OpBuilder`:
+To effectively interface with MLIR, we will define a new Toy dialect. This
+dialect will properly model the semantics of the Toy language, as well as
+provide an easy avenue for high-level analysis and transformation.
 
--   The name of the operation.
--   A location for debugging purposes. It is mandatory, but can be explicitly
-    set to `unknown`.
--   A list of operand values.
--   A list of types for result values.
--   A list of attributes.
--   A list of successors blocks (for branches mostly).
--   A list of regions (for structural operations like functions).
+```c++
+/// This is the definition of the Toy dialect. A dialect inherits from
+/// mlir::Dialect and registers custom attributes, operations, and types (in its
+/// constructor). It can also override some general behavior exposed via virtual
+/// methods.
+class ToyDialect : public mlir::Dialect {
+ public:
+  explicit ToyDialect(mlir::MLIRContext *ctx);
+
+  /// Provide a utility accessor to the dialect namespace. This is used by
+  /// several utilities for casting between dialects.
```

Comment:
```suggestion
  /// several utilities.
```

I'm not sure what "casting between dialects" means here, and there are likely other uses so we can just drop I think.